### PR TITLE
feat(recap): Tweaks the process_recap_email task to ignore sealed entries

### LINKED ISSUE #2835: @recap.email does not avoid sealed docket entries (it only does sealed documents)
Right now, when @recap.email gets a NEF about a sealed document, we can identify that by trying to download the document. If we successfully get it, we know it's not sealed, and if we can't get it, it is. Simple enough.

In [this docket](https://www.courtlistener.com/docket/6923200/mwk-recruiting-inc-v-jowers/?order_by=desc), you can see that there is no public entry for number 399. This is *not* because the entry doesn't exist. It's because the entry *itself* is sealed, as well as the document. 

When this is the case:

1. There is a gap in the numbering of the public docket. For example, if docket entry 3 is sealed, the docket will show numbers 1 and 2, skip number 3, and immediately show number 4.

2. The attachment page will fail as well.

Our error in this instance was to parse the NEF and to add the entry to CL while marking the document as sealed. Instead, we should have figured out that the entry was sealed, and added nothing to the site.

To fix this issue:

1. When we identify that a document is sealed...
2. We should check if the attachment page for it is available.
    - If so, we should not add anything to our DB.
    - If not, then we should do what we do now (add the entry, but mark the doc as sealed)

What else?

### PR DESCRIPTION
**Pull request by ERosendo**
This PR enhances the `process_recap_email` task by adding logic to check potential sealed entries in district docket reports. Key changes include:

- Introduces a helper named `is_docket_entry_sealed`: This method uses Juriscraper to analyze the download confirmation page and determine if an entry is likely sealed.
 
- Enhances `process_recap_email` to prevent sealed entry creation:
   - If the recap email lacks attachments and the entry is flagged as potentially sealed, no new entry is created.
   - If the recap email has attachments and the entry seems sealed:
      - We attempt to access the attachment page and extract any available documents.
      - If accessing the attachment page fails, the newly created docket entry is removed.

This PR fixes #2835

### CODE REVIEW

**Review by albertisfu — COMMENTED** [review]
Great, the general architecture seems right to me!

However, I have a couple of questions.

I reviewed some of the sealed docket entries examples you attached in #2835. There are some attachments that are sealed, and their receipt page shows the message: `"could not retrieve dktentry..."` For example:

Like: https://www.courtlistener.com/docket/64947405/doe-v-mast/#entry-171
Entry 171 attachment 8: https://ecf.vawd.uscourts.gov/doc1/19114615670?caseid=126284

Checking the attachment page for the entry:
https://ecf.vawd.uscourts.gov/doc1/19104615662

In this case, it is available.

Screenshot 2024-02-15 at 16 50 13

However, attachment 8 is not published on the attachment page, while in CL it is published but marked as sealed.

Screenshot 2024-02-15 at 16 54 38


So two questions:

1. The new Juriscraper method **`is_entry_sealed`** checks the main document URL using the main **`pacer_doc_id`** and looks for the `"could not retrieve dktentry..."` message to determine whether the entry is sealed. During the process of merging attachments, you check again if the attachment page is available to determine if the entry should be removed or not. If I'm not missing something, the receipt page we query in **`is_entry_sealed`** and the attachment page use the same URL since both are based on the main **`pacer_doc_id`**, correct? If that's the case, I think we could just abort the process of adding the entries if **`is_potentially_sealed_entry`** is `True` at the beginning without worrying about whether the entry has attachments or not.

2. How should we approach the case described above, where the main entry is not sealed so the attachment page is available, but one of the documents is sealed and shows the message `"could not retrieve dktentry...",` and is not published on the attachment page on PACER? Should we also hide or remove the RECAPDocument in CL?

**Comment by mlissner** [comment]
I'm having a bit of trouble following the design without reading the code, but I'll share a couple things that I think will clarify. If they don't, tell me, and I'll go read the code:

1. If the docket entry is sealed, then we can assume all of the documents and attachments in the entry are also sealed.

2. When there are attachments in an unsealed docket entry:
    - The main document alone can be sealed while the attachments are not.
    - The attachments can be sealed while the main document is not.
    - Everything can be sealed.
    - Nothing can be sealed.
    - In other words, any combination is possible. :)

But I think 1 is the important point that might help?

**Comment by ERosendo** [comment]
@mlissner thanks for your input regarding this issue.

> I think 1 is the important point that might help?

Yeah, I was hesitant about skipping attachment checks because there were only a few instances of it in the dataset. However, I tweaked the code of the **email processing task** to bypass calls to helper methods for attachment checks if the entry is identified as sealed. Additionally, I've updated the tests to ensure they reflect this new behavior.

File: cl/recap/tests.py [hunk 1]
```diff
@@ -4434,6 +4463,169 @@ async def test_recap_email_minute_entry(
...
+        pq_query = ProcessingQueue.objects.filter(
+            pacer_doc_id=docket_entry[0]["pacer_doc_id"],
+            pacer_case_id=docket_entry[0]["pacer_case_id"],
+        )
+        self.assertEqual(await pq_query.acount(), 1)
+        processing_queue = await pq_query.afirst()
+        self.assertIn(
+            "Failed to get docket entry", processing_queue.error_message
+        )
+
```
Comment (albertisfu) [inline cl/recap/tests.py:4533]: I think we should include the following assertion at this point:
`self.assertEqual(len(mail.outbox), 0)`

With the current logic, alerts are not triggered because entries are not created; however, if we alter the logic in the future to create docket entries and merely mark them as sealed, we should ensure that alerts are not triggered for these sealed entries.

File: cl/recap/tests.py [hunk 2]
```diff
@@ -4434,6 +4463,169 @@ async def test_recap_email_minute_entry(
...
```
Comment (albertisfu) [inline cl/recap/tests.py:4620]: Similar as the previous comment:
`self.assertEqual(len(mail.outbox), 0)`

**Review by albertisfu — COMMENTED** [review]
Changes look great!

I just left one suggestion for the tests.

- Additionally, the `WebhooksRetries` test class also processes emails, so we should include the `is_docket_entry_sealed` mock as well.

- Just to confirm, about my [comment ](https://github.com/freelawproject/courtlistener/pull/3788#pullrequestreview-1884049894), It seems that an attachment document can be sealed and still shown on the attachment page. However, it also appears that sometimes the attachment entry itself is sealed and not displayed on the attachment page. If we query the receipt page for that document, we'll receive a "docket entry sealed" message. Should we also avoid showing those attachment entries on the docket page?

- Finally, one question related to PACER, just to be sure the logic used is correct. In a multi-docket NEF, we receive the same entry for multiple cases. In those instances, we added the entry to each docket. Is it possible for the entry to be sealed in one case while it is available in others, considering each case has a different `pacer_case_id`?

File: cl/recap/tests.py (hunk 1 above)
Comment (ERosendo) [reply cl/recap/tests.py:4533]: Thanks for your suggestion. I'll apply this change.

**Comment by ERosendo** [comment]
@albertisfu thanks for your review. I've applied your suggestions. 

> However, it also appears that sometimes the attachment entry itself is sealed and not displayed on the attachment page. If we query the receipt page for that document, we'll receive a "docket entry sealed" message. Should we also avoid showing those attachment entries on the docket page?

I thought about your comment and I believe we can find two possible scenarios with sealed documents:

1. If the document is sealed and not listed on the attachment page, it won't be created in CL and there won't be an issue.
2. the document is listed on the attachment page, but we receive a "docket entry sealed" message when trying to access it. In this case, instead of skipping the record creation (which is currently the only way we have to hide the document), we should mark the document as sealed. This way, we can differentiate it from other documents and prevent recap uploads from adding the docket back to the entry. 

Automating sealing documents using the `is_sealed` flag during `recap_email_processing` removes the need for manual flagging in the future, saving time and effort.

I checked the source of the example entry you provided and found out that the documents which are no longer listed were once available There are six processing queues associated with this attachment page, and we've even received a PDF upload for document 8 (which is no longer listed). Here are some of the htmlfiles:

- https://www.courtlistener.com/admin/recap/pacerhtmlfiles/5939567/change/
- https://www.courtlistener.com/admin/recap/pacerhtmlfiles/5947826/change/
- https://www.courtlistener.com/admin/recap/pacerhtmlfiles/5951204/change/

**Comment by albertisfu** [comment]
Got it, thanks for the changes and the detailed explanation.

Yes, it makes sense that Attachment #8 was published at the time the attachment page was parsed, but the document was sealed. Afterwards, the attachment was also removed from the attachment page on PACER.

We already check for attachment documents being sealed when processing an email, which seems sufficient for now since we don't have a field for marking `RECAPDocuments` as not published. If we avoid creating the RECAPDocument as you described, once someone else uploads the attachment document from a different source, it will be created again, not being sealed at all. Which is what happened with the docket entry used as an example in this issue. So, a future improvement, as we discussed, is to add the sealed entry field to `DocketEntry`.

This seems good to me now. Just if @mlissner  could help us to confirm this is not an issue, I think we are ready.

> Finally, one question related to PACER, just to be sure the logic used is correct. In a multi-docket NEF, we receive the same entry for multiple cases. In those instances, we added the entry to each docket. Is it possible for the entry to be sealed in one case while it is available in others, considering each case has a different pacer_case_id?

**Review by albertisfu — APPROVED** [review]

**Comment by mlissner** [comment]
Thanks for discussing and figuring out these issues. You asked:

>  Is it possible for the entry to be sealed in one case while it is available in others, considering each case has a different pacer_case_id?

I guess anything is possible, but it'd be weird semantically. Like, it wouldn't make sense to seal in one case, but not the other. Does that mean we need more work here or that we're good to go?

**Comment by mlissner** [comment]
This looks good enough to me, so I'm merging it. Two remaining pieces:

1. The question above: Does it require more work?
2. Eduardo, you identified some content that we should seal in prod, right? Do you have that list and maybe a script I can run to seal it?

**Comment by albertisfu** [comment]
> I guess anything is possible, but it'd be weird semantically. Like, it wouldn't make sense to seal in one case, but not the other. Does that mean we need more work here or that we're good to go?

Got it. Yeah, here we're checking if the main entry is sealed. Only if we find a case where one entry is sealed and other isn't  in a multi-docket NEF, the logic will need a tweak.

**Comment by mlissner** [comment]
But in the meantime, we handle that by defaulting to sealed for all items if the first one is sealed, right?

**Comment by albertisfu** [comment]
But in the meantime, we handle that by defaulting to sealed for all items if the first one is sealed, right?

Yes, this is correct. However, we don't "seal" entries. The logic in this code prevents merging the entries for both the first and all other dockets if the first entry is found to be sealed.
